File: 02 Leet - Medium/48-rotate-image.js
/*

You are given an n x n 2D matrix representing an image.

Rotate the image by 90 degrees (clockwise).

Note: You have to rotate the image in-place, which means you have to modify the input 2D matrix directly. DO NOT allocate another 2D matrix and do the rotation.

*/

let matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];

let rotate = matrix => {
  // Assume N x N matrix
  let mLen = matrix.length;

  // Transpose matrix
  for (let i = 0; i < mLen; i++) {
    for (let j = i; j < mLen; j++) {
      [matrix[i][j], matrix[j][i]] = [matrix[j][i], matrix[i][j]];
    }
  }

  // Reverse each row
  for (let i = 0; i < mLen; i++) {
    matrix[i] = matrix[i].reverse();
  }

  return matrix;
};

// Test
console.log(rotate(matrix));

File: 02 - Medium/48-rotate-image.py
'''
You are given an n x n 2D matrix representing an image.

Rotate the image by 90 degrees (clockwise).

Note: You have to rotate the image in-place, which means you have to modify the input 2D matrix directly. DO NOT allocate another 2D matrix and do the rotation.
'''


class Solution:
    def rotate(self, matrix):
        for i in range(len(matrix)):
            for j in range(i, len(matrix)):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

        for i in range(len(matrix)):
            matrix[i] = matrix[i][::-1]

        return matrix


matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]
sol = Solution()
print(sol.rotate(matrix))
